### DISCUSSION TITLE
Process fails to run rsync, copying command to CLI works fine?

### DISCUSSION THREAD

**Discussion by letharion** [symfony/symfony#45078] [Q&A] [answered]
I have the following code:

    $executableFinder = new ExecutableFinder();
    $sshBin = $executableFinder->find('ssh');
    $rsync = [
      'rsync',
      "-e '$sshBin -p $remote_backup_server_port'",
      "$remote_backup_server_user@$remote_backup_server:$remote_backup_path/$remote_db_backup_name",
      '.',
    ];
    $process = new Process($rsync);
  
Which worked at one point, but now throws:

    rsync: [Receiver] Failed to exec /usr/bin/ssh -p 3022: No such file or directory (2)
    rsync error: error in IPC code (code 14) at pipe.c(85) [Receiver=3.2.3]
    rsync: connection unexpectedly closed (0 bytes received so far) [Receiver]
    rsync error: error in IPC code (code 14) at io.c(228) [Receiver=3.2.3]

Initially I had just `ssh` as the command and figured it was due to that, but letting the ExecutableFinder dig out the absolute path didn't help.
If I var_dump($rsync) I can copy paste the command to the CLI and it works fine.

Any ideas?

**Comment by stof** [comment, marked as answer]
The issue to me is your `"-e '$sshBin -p $remote_backup_server_port'"` string in the array passed to the Process. You don't want to pass a single argument containing `-e '/usr/bin/ssh -p 3022` in the command line. You want to pass `-e` and then `/usr/bin/ssh -p 3022` as the following value.

```php
$p = new Process([
  'rsync',
  '-e'
  "$sshBin -p $remote_backup_server_port",
  "$remote_backup_server_user@$remote_backup_server:$remote_backup_path/$remote_db_backup_name",
  '.',
]);
// running `rsync -e "/usr/bin/ssh -p 3022" "foo@bar:baz/db" .`

$p2 = new Process([
  'rsync',
  "-e '$sshBin -p $remote_backup_server_port'",
  "$remote_backup_server_user@$remote_backup_server:$remote_backup_path/$remote_db_backup_name",
  '.',
]);
// running `rsync "-e '/usr/bin/ssh -p 3022'" "foo@bar:baz/db" .`
```

**Reply by stof** [reply]
if you have a command that you run in your terminal, creating the array for the Process class means putting each space-separated argument into a separate item of the array (and for those items, don't include the terminal-level quoting or escaping of values, as that will be handled by the Process class)

**Comment by letharion** [comment]
I would have absolutely sworn that it did work initially, but clearly the evidence says otherwise.
Thanks, I've broken down the command to put the ssh command in a separate argument, and it works fine now. :)
